Adapted ESRF codebase to mxcubecore

marcus-oscarsson opened the pull request:
This PR contains a few general fixes for mxcubcore

 - A socket module patch for Py4J (provided by @rhfogh)
 - Integration with GPHL workflows (provided by @rhfogh and @olofsvensson)
 - A few other minor fixe

It also contains the changes needed to get the ESRF specific HardwareObjects and routines
to be compatible with mxcubecore.

rhfogh commented on mxcubecore/HardwareObjects/MicrodiffZoom.py:
```diff
@@ -43,9 +43,12 @@
     def init(self):
         """Initialize the zoom"""
-        super().init()
+        ExporterNState.init(self)
```
Why not super()? Either should work, of course, but super() seems cleaner

rhfogh commented on mxcubecore/HardwareObjects/XMLRPCServer.py:
```diff
@@ -672,3 +680,32 @@ def _register_module_functions(self, module_name, recurse=True, prefix=""):
+
+    def addXrayCentring(self, parent_node_id, **centring_parameters):
+        """Add Xray centring to queue.
+        NB can also be accessed as a static function of XMLRPCServer
```
The comment about static functions are obsolete and should be deleted

rhfogh commented on mxcubecore/HardwareObjects/XMLRPCServer.py:
```diff
@@ -672,3 +680,32 @@ def _register_module_functions(self, module_name, recurse=True, prefix=""):
+
+    def addGphlWorkflow(self, parent_node_id, task_dict):
+        """Add GPhL owrkflow to queue.
+        NB can also be accessed as a static function of XMLRPCServer
```
The comment about static functions are obsolete and should be deleted

rhfogh commented on mxcubecore/HardwareObjects/queue_entry.py:
```diff
@@ -470,6 +470,17 @@ def __init__(self, view=None, data_model=None):
         BaseQueueEntry.__init__(self, view, data_model)
 
 
+class DummyQueueEntry(BaseQueueEntry):
```
Rename to DelayQueueEntry is probably better, as you suggested.

rhfogh commented on mxcubecore/HardwareObjects/queue_entry.py:
```diff
@@ -1853,13 +1924,15 @@ def center_before_collect(view, dm, queue, sample_view):
     queue_model_objects.TaskGroup: TaskGroupQueueEntry,
     queue_model_objects.Workflow: GenericWorkflowQueueEntry,
+    queue_model_objects.XrayCentring2: XrayCentring2QueueEntry,
     queue_model_objects.XrayCentering: XrayCenteringQueueEntry,
```
Should eventually be refacto0red away, but this is probably not the time.

marcus-oscarsson replied:
I agree

rhfogh commented on mxcubecore/HardwareObjects/queue_model_objects.py:
```diff
@@ -30,7 +30,10 @@
 from mxcubecore.HardwareObjects import queue_model_enumerables
 
-from mxcubecore import HardwareRepository as HWR
+try:
```
That is indeed in the file, but why do we want to continue quietly if HardwareRepository does not load? HWR is used not only in GphlWorkflow, but also in XrayCentring2 and in Sample

marcus-oscarsson replied:
Indeed, queue_model_objects was originally written to be self contained so that it could be 
easily transferred without having to send or have all of mxcube. However the introduction of
`from mxcubecore import HardwareRepository as HWR` obviously breaks this.

So we need to add this try catch inorder for the BES workflows to work,

rhfogh replied:
HWR is used extensively in GphlWorkflow, but that is OK - that is a special and optional part of the whole. But HWR is also used in XrayCentring2 and Sample, which might trip somebody up. Is there a way of getting to HWR.beamline.session that does not require HWR to load, so we can get rid of HWR from those classes? What is the kind of thing that might go wrong to break the import but would still allow BES to work - just for my education?

olofsvensson replied:
Yes, I think moving GphlWorkflow out of QMO - as well as XrayCentring2 and Sample - will solve the problem.  It would then be clearer that all objects in QMO would not depend on HWR or any other modules - as Marcus said, the idea was that the file should be self contained. The question is then where GphlWorkflow, XrayCentring2 and Sample would go?

marcus-oscarsson replied:
It does not break anything at the moment because neither Sample nor XrayCenterint2 is used from BES directly (as far as I know). 

We have already talked about dividing queue_model_objects in smaller pieces and I have also started to look at all this in #686. I would suggest that we keep this as it as for the moment, and focus on reworking this in #686.

rhfogh replied:
GphlWorkflow I can put in the Gphl subdirectory, no problem - GPhL workflows are a bit of a niche product, and requires bespoke set-up one way or the other. But if something as basic as Sample is moved out, that would seem to raise the question which ModelObjects are supposed to be in the QMO file and which not. Which ModelObjects must be available without connection to the rest of MXCuBE and for which purposes? I was never aware that there was a limitation on what QMOs could import, so maybe I had better find out before I make furtehr code cahnges?

marcus-oscarsson replied:
It's indeed something that we overlooked when we started to use the HWR import. Its alright as it is for the moment, but we should not extend the usage of HWR until we have found a way to handle this. As I mentioned above I think the right place to do so is in #686 or maybe even in a separate PR. 

I can think of several solutions that we could discuss, to work around this issue.

We could simply install mxcubcore on the BES side, in worst case scenario as we are no longer really transferring queue_model_obejects like we used to do.

rhfogh commented on mxcubecore/HardwareObjects/queue_model_objects.py:
```diff
@@ -2004,25 +2159,30 @@ def set_pre_strategy_params(
             orgxy = HWR.beamline.detector.get_beam_position(
                 distance, wavelength
             )
+            orgxy = tuple(x / 0.172 for x in orgxy)
```
Should be removed, assuming get_beam_position now returns pixels

beteva commented on mxcubecore/HardwareObjects/ExporterNState.py:
```diff
@@ -45,15 +45,15 @@ class ExporterNState(AbstractNState):
 
     def __init__(self, name):
-        AbstractNState.__init__(self, name)
+        super.__init__(name)
```
super() - missing brackets?

marcus-oscarsson replied:
missing brackets ?

beteva replied:
I believe it is `super().__init__(name)` and not `super.__init__(name)`

beteva commented on mxcubecore/HardwareObjects/ExporterNState.py:
```diff
@@ -45,15 +45,15 @@ class ExporterNState(AbstractNState):
     def init(self):
         """Initialise the device"""
-        AbstractNState.init(self)
+        super.init()
```
super() - missing brackets?

rhfogh reviewed (changes requested):
Very good indeed! Approved, just with a few minor points.

marcus-oscarsson commented on mxcubecore/configuration/esrf_id30a2/gphl_beamline_config/transcal.nml:
```diff
@@ -0,0 +1,5 @@
+&SDCP_INSTRUMENT_LIST
```
I imagine that you will make a complementary PR with final changes for your workflows. I'll leave to you to look over your configuration files in greater detail. I will simply remove this file

rhfogh replied:
I will. Just for checking: Is configuration/esrf_id30a2/gphl_beamline_config exactly the configuration files used at MASSIF1, or are the actual configuration files different and living somewhere  else? Since responsibility for maintaining them is to some extent shared it can get confusing.

marcus-oscarsson replied:
Oh, yes its the ones from MASSIF1

marcus-oscarsson commented:
Thanks for having a look and being so responsive @beteva, @rhfogh and @olofsvensson. I think I've addressed the comments above. I will let you @rhfogh follow up with a PR related to your internal workflow things, like configuration files and so on. I'm sure there are also some other improvements you would like to add after the next test.
